draft/keymanagement.md

```diff
@@ -0,0 +1,129 @@
+- `id`
+- `addr` - address of the contact which is supposed to have the private key
+- `public_key` - a blob with the OpenPGP public key
+- `public_key_fingerprint` - public key fingerprint
+- `introduced_by` - address of the contact if received directly in `Autocrypt` header, address of `Autocrypt-Gossip` sender otherwise
+- `is_verified` - boolean flag indicating if the key was received in a verified group or not
+- `timestamp` - timestamp of the most recent row update 
+
+The table has consraints `UNIQUE(addr, introduced_by)`.
+Note that there may be multiple rows with the same fingerprint,
+e.g. if multiple peers gossiped the same key for a contact they all are introducers of the key.
+
+`prefer_encrypted` is moved to the `contacts` table.
+
+
+## Using the table to select the keys
+
+### Sending a message to 1:1 chat
+
+When sending a message to the contact in a 1:1 chat,
+encrypt to the key where `introduced_by` equals `addr`:
+```
+SELECT *
+FROM public_keys
+WHERE addr=? AND introduced_by=addr
+```
+This row is guaranteed to be unique if it exists.
+
+If direct Autocrypt key does not exist,
+use the most recent one according to the timestamp key gossiped for this contact:
+```
+SELECT *
+FROM public_keys
+WHERE addr=?
+ORDER BY timestamp DESC LIMIT 1
+```
+
+If the table contains a row with selected key and `is_verified` flag set,
+display a green checkmark and "Introduced by <introduced_by>" in the 1:1 chat/contact profile.
+Note that row with verification may be older than the most recent gossip.
+
+### Sending a message to unprotected group chat
+
+When sending a message to the contact in a non-verified group chat,
+encrypt to the same key as used in 1:1 chat
+and (NEW!) one or more most recent gossiped keys introduced by chat members:
+```
+SELECT public_key
+FROM public_keys
+WHERE addr=? AND introduced_by IN (<chat-member-list>)
+ORDER BY timestamp DESC
+```
+It does not matter if the gossiped key has `is_verified` flag.
+
+### Sending a message to protected group chat
+
+When sending a message in a protected group chat,
+construct a list of candidate keys for each contact
+the same way as in an unprotected chat,
+but then filter out the keys which do not have `is_verified` flag.
+
+
+## Updating the table
+
+When executing "setup contact" protocol
+with the contact,
+set the row with `addr` and `introduced_by` being equal
+to the contact address and `is_verified` is true.
+In this case the contact becomes directly verified.
+
+When receiving a message,
+first process the `Autocrypt` key,
+then check verification properties to detect if gossip is signed with a verified key,
+then process `Autocrypt-Gossip` key.
+
+### 1. Processing the Autocrypt header
+
+Take the key from the Autocrypt header
+and insert or update the row
+identified by `addr` and `introduced_by` being equal to the `From:` address.
+Notify about setup change and reset `is_verified` flag if the key fingerprint has changed.
+Always update the timestamp.
```

Comment:
This seems to mean that we can't encrypt to a green group anymore if we receive a changed fingerprint from a group member in a different chat?

As I understand this paragraph, per one email address, there will always be only one row with a direct key. And  "if the key fingerprint has changed" (so in the scenario above, if we receive a message in a different chat with a new unverified key), no new row is created, but the existing one is updated with the new fingerprint + public key, so the old key is lost. This is what we want in normal Autocrypt mode.

But if we have a green group with the contact, which was established by scanning their QR code directly, we would lose the verified public key we used to encrypt in that chat. As I understand the `### Sending to a protected group chat` paragraph, the query to get the contact's key would be:

```
SELECT public_key
FROM public_keys
WHERE addr=? AND is_verified=True AND introduced_by IN (<chat-member-list>)
ORDER BY timestamp DESC
```

This SQL statement would return no rows, until someone else in that group sends us a verified gossip header with the old verified key in it. So we could not send any messages there.

Reply:
I'd rather say that timestamp shouldn't be updated from gossips like it's done for the Autocrypt header, i.e. from the Date header. For a gossip there should be a separate field in the Autocrypt-Gossip header which is derived from the sender's `public_keys` table. As for the problem with encryption, maybe just not encrypt to the member whose key was changed to a new unverified one. It still has sense to send messages to other group members.

> Then we need to be able to have both a "verified" and non-verified key and add `is_verified` into the UNIQUE constraint.

But anyway this is useful, otherwise we lose the existing verified key in case of an attack on Autocrypt or the contact's email account theft. And also can't verify previously received messages anymore.